AK/Result.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <AK/Assertions.h>
+#include <AK/Platform.h>
+#include <AK/Optional.h>
+
+namespace AK {
+
+template<typename T, typename E>
+class CONSUMABLE(unknown) Result {
+public:
+    RETURN_TYPESTATE(unknown)
+    Result(const T& res)
+        : m_result(res)
+    {}
+
+    RETURN_TYPESTATE(unknown)
+    Result(const E& error)
+        : m_error(error)
+    {
+    }
+
+    RETURN_TYPESTATE(unknown)
+    Result(const T& res, const E& error)
+        : m_result(res)
+        , m_error(error)
+    {
+    }
+
+    RETURN_TYPESTATE(unknown)
+    Result(Result&& other)
+        : m_result(move(other.m_result))
+        , m_error(move(other.m_error))
+    {
+    }
+
+    RETURN_TYPESTATE(unknown)
```

Comment:
I think that allowing an unknown Result to destruct makes sense. But does it?

Why would you want to write:

f.read(); // and ignore either the error or the result?

That seems a little weird, but maybe it's OK. Imagining e.g. close() when you're about to exit.